### ISSUE TITLE
mx.io.ImageRecordIter confused me, does this function is related to time?

### ISSUE THREAD

**Issue by chenboheng** [apache/incubator-mxnet#7708]
I use the latest mxnet code (0.11), and I want to study the usage of the mx.io.ImageRecordIter.
I use a simple image dataset which includes 16 images. This small dataset is downloaded from http://data.mxnet.io/data/test_images.tar.gz. I convert this image database as testim.rec

If I run the following code:

import mxnet as mx
import matplotlib.pyplot as plt
import time
import numpy as np

def get_iterators(datapath,batch_size,data_shape=(3,448,448)):
    dataiter = mx.io.ImageRecordIter(
        path_imgrec         = datapath,
        batch_size          = batch_size,
        data_shape          = data_shape,
        )
    return dataiter

batch_size=6
datapath='data/testim.rec'
dataiter = get_iterators(datapath, batch_size)
#time.sleep(2)
dataiter.reset()

number=0
for batch in dataiter:
    data=batch.data[0]
    for i in range(batch_size):
        plt.subplot(3,batch_size, i+1+number*batch_size)
        plt.imshow(data[i].asnumpy().astype(np.uint8).transpose((1, 2, 0)))
    number = number + 1

plt.show()

The result will be:
![figure_1](https://user-images.githubusercontent.com/20148944/29994319-dfa910f2-8fff-11e7-9718-c1c239b3186a.png)

If I run the following code:

import mxnet as mx
import matplotlib.pyplot as plt
import time
import numpy as np

def get_iterators(datapath,batch_size,data_shape=(3,448,448)):
    dataiter = mx.io.ImageRecordIter(
        path_imgrec         = datapath,
        batch_size          = batch_size,
        data_shape          = data_shape,
        )
    return dataiter

batch_size=6
datapath='data/testim.rec'
dataiter = get_iterators(datapath, batch_size)
time.sleep(2)
dataiter.reset()

number=0
for batch in dataiter:
    data=batch.data[0]
    for i in range(batch_size):
        plt.subplot(3,batch_size, i+1+number*batch_size)
        plt.imshow(data[i].asnumpy().astype(np.uint8).transpose((1, 2, 0)))
    number = number + 1

plt.show()

The result will be:
![figure_2](https://user-images.githubusercontent.com/20148944/29994331-05b9892a-9000-11e7-9827-78a7f83395a7.png)

The second code only sleep two seconds before dataiter.reset(). The two iteration results are not same. Does the iterator is related to the time? Or I write wrong codes?

**Comment by chenboheng** [comment]
I use mx.image.ImageIter to replace mx.io.ImageRecordIter. The two codes obtain the same results. The usage of mx.io.ImageRecordIter confuses me.

**Comment by ptrendx** [comment]
This seems like a bug - I will look at it tomorrow.

**Comment by ptrendx** [comment]
Ok, I did reproduce it. Will debug it now.

**Comment by ptrendx** [comment]
Ok, I understand why it happens:
reset() function in Python calls BeforeFirst function in C interface, which is supposed to reset the iterator. BeforeFirst function is also used after the epoch ends to reset the iterator. Since the dataset size may be indivisible by the batch size (and it is in the example shown by @chenboheng), by default ImageRecordIter resets itself and takes images from the next epoch to fill the last batch (it is shown in the first output shown in this issue - 2 last images from the last batch are the same as 2 first images of the first batch). Then the "reset" of iterator does not actually reset it - if it was, it would read those 2 images twice. 
In order to improve performance, ImageRecordIter prefetches a few batches of data. It turns out that this dataset is small enough that the iterator prefetches all of it in those 2 seconds when the main thread sleeps, and so it reaches the end of epoch internally. Calling reset in that situation calls BeforeFirst in the same way it would be called after epoch is done, so it does not actually reset the iterator, but prepares it for a new epoch, with dataset being offset by those 2 images that it used to fill the last batch of the "previous epoch". That is why the result is different.  
So basically the bug can only be observed if the dataset is small enough that it fits in internal prefetch queue, the size of dataset is not divisible by the batch size and we give it enough time to internally process the entire dataset.

I guess the possible solution would be to treat the BeforeFirst as the end of epoch and not hard reset only if the end of epoch was actually reached through calls to next. I will look into fixing this. 
Still, the "reset" name kind of suggests it will completely reset the iterator and that is not the case if you reach end of epoch. @piiswrong What do you think about it? Is it acceptable?

**Comment by ptrendx** [comment]
After looking into possible solutions I think we can close it without fixing - the possible fix is pretty complicated and touches much more than just the iterator, the behavior shown in the issue is a corner case not used in practice and it does not actually affect training results (data is just offset by a few images). I guess we should document that behavior though. 
@piiswrong please close.

**Comment by ptrendx** [comment]
For @chenboheng - one thing to note is that you do not need to call reset() after creating iterator - it is already reset.

**Comment by chenboheng** [comment]
@ptrendx. I will not call reset() after creating iterator. Thanks a lot

**Comment by ptrendx** [comment]
@szha please close
